yt/frontends/ramses/fields.py

```diff
@@ -241,35 +256,83 @@ def create_cooling_fields(self):
                      'logT' : np.log10(data["temperature"]).ravel()}
-                rv = 10**interp_object(d).reshape(shape)
-                # Return array in unit 'per volume' consistently with line below
-                return data.ds.arr(rv, number_density_unit)
+                rv = interp_object(d).reshape(shape)
+                if not(name[-1]=='mu'):
+                    rv = 10**interp_object(d).reshape(shape)
+                cool=data.ds.arr(rv, unit)
+                if 'metal' in name[-1].split('_'):
+                    cool=cool*data['metallicity']/0.02  #Ramses uses Zsolar=0.02
+                elif 'compton' in name[-1].split('_'):
+                    cool=data.ds.arr(rv, unit+'/cm**3')
+                    cool=cool/data['number_density']    #Compton cooling/heating is written to file in erg/s
+                return cool
             self.add_field(name=name, sampling_type="cell", function=_func,
-                           units=self.ds.unit_system['number_density'])
+                           units=unit)
+
+        #Load cooling files
         avals = {}
         tvals = {}
         with FortranFile(filename) as fd:
             n1, n2 = fd.read_vector('i')
-            n = n1 * n2
             for ax in _cool_axes:
                 avals[ax] = fd.read_vector('d')
-            for tname in _cool_arrs:
+            for i,(tname, unit) in enumerate(_cool_arrs):
                 var = fd.read_vector('d')
+                if var.size==n1 and i==0:
+                    #If this case occurs, the cooling files were produced pre-2010 in a format
+                    #that is no longer supported
+                    return
                 if var.size == n1*n2:
-                    tvals[tname] = var.reshape((n1, n2), order='F')
+                    tvals[tname] = dict(data=var.reshape((n1, n2), order='F'), unit=unit)
                 else:
                     var = var.reshape((n1, n2, var.size // (n1*n2)), order='F')
                     for i in range(var.shape[-1]):
-                        tvals[_cool_species[i]] = var[:,:,i]
+                        tvals[_cool_species[i]] = dict(data=var[:,:,i], unit="1/cm**3")
+
+        #Add the mu field first, as it is needed for the number density
+        interp = BilinearFieldInterpolator(tvals['mu']['data'],
+                                           (avals["lognH"], avals["logT"]),
+                                           ["lognH", "logT"], truncate = True)
+        _create_field(("gas", 'mu'), interp, tvals['mu']['unit'])
+
+        #Add the number density field, based on mu
+        def _number_density(field,data):
+            return data[('gas','density')]/mp/data['mu']
+        self.add_field(name=('gas','number_density'),sampling_type="cell",function=_number_density,
+                       units=number_density_unit)
+
+        #Add the cooling and heating fields, which need the number density field
+        for key in tvals:
+            if not(key=='mu'):
```

Comment:
I'd rather use `if key is not 'mu'` as I find it a bit more clear but that's a personal preference.